### PR TITLE
Log to both logjam and a logger at specifiable levels

### PR DESCRIPTION
**Pull request by skaes**
Adds a new package called "log" which is an extension of the standard go logger than can send logs to both logjam and a file. Can be configured with different logging thresholds for each logger.

### CODE REVIEW

**Comment by skaes** [comment]
@roccoblues I'm wondering whether we should drop the Log* methods from the logjam package?

**Comment by skaes** [comment]
Dropped the log package.

File: agent.go [hunk 1]
```diff
@@ -39,6 +43,10 @@ var agent struct {
 	topic     string      // The default log topic
 }
 
+func init() {
```
Comment (roccoblues) [inline agent.go:46]: leftover? Or is this called somewhere?

File: agent.go [hunk 2]
```diff
@@ -50,7 +58,8 @@ type Options struct {
 	Rcvhwm              int                 // ZeroMQ socket option of the same name
 	Sndtimeo            int                 // ZeroMQ socket option of the same name
 	Rcvtimeo            int                 // ZeroMQ socket option of the same name
-	Logger              Logger              // Logjam errors are printed using this interface.
+	Logger              Printer             // Logjam errors are printed using this interface.
```
Comment (roccoblues) [inline agent.go:61]: Seeing this two so close together first implied some kind of connection to me. Even though the description is pretty clear. What do you think about renaming `Logger` to `ErrorLogger`?

File: middleware.go [hunk 3]
```diff
@@ -45,7 +45,10 @@ func (m *middleware) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 
 	defer func() {
 		if recovered := recover(); recovered != nil {
-			logjamRequest.Log(FATAL, fmt.Sprintf("%#v", recovered))
+			msg := fmt.Sprintf("%#v", recovered)
+			agent.Logger.Println(msg)
```
Comment (roccoblues) [inline middleware.go:49]: Is this really a logjam error that we should log?

File: util.go [hunk 4]
```diff
@@ -16,36 +17,92 @@ const (
...
+	l.log(ctx, ERROR, args...)
 }
 
-// LogError calls Log with ERROR severity.
-func LogError(ctx context.Context, format string, args ...interface{}) {
-	Log(ctx, ERROR, format, args...)
+// FatalPanicf logs with FATAL severity, then panics. Please note that the standard Logger
+// method exits the program, which is usually not a desired outcome in a long running
+// server application.
+func (l *Logger) FatalPanicf(ctx context.Context, format string, args ...interface{}) {
```
Comment (roccoblues) [inline util.go:97]: I don't like the `FatalPanic` methods and would say we don't need them. `Fatal` in my eyes is an unrecoverable error that should bubble up and be catched by a `recover` or terminate the program. If I see a `FATAL` logline in logjam I assume that my request failed and the error was caught and logged by logjam.

**Review by roccoblues — COMMENTED** [review]
This is really cool! I only have a some small suggestions and would argue for removing the `LogFatal*` methods. See my comment inline

File: agent.go (hunk 2 above)
Comment (skaes) [reply agent.go:61]: The package needs a logger when running tests. If we don't do this or something similar, then users will need to set up the logger in their tests. Which comes with new problems: when and were to do and we would need to document it.

File: agent.go (hunk 1 above)
Comment (skaes) [reply agent.go:46]: The package needs a logger when running tests. If we don't do this or something similar, then users will need to set up the logger in their tests. Which comes with new problems: when and were to do and we would need to document it.

File: agent.go (hunk 2 above)
Comment (skaes) [reply agent.go:61]: I'd like to stay vague. We might want to log more than just errors in the future.

File: util.go (hunk 4 above)
Comment (skaes) [reply util.go:97]: I would like to have a method that can log at log level fatal, but this is already taken by the standard logger methods. And the Fatal method of the standard logger calls os.Exit, so nothing bubbles up. We could make it more explicit what happens by renaming it to LogFatalAndPanicf, but I don't want to drop it.

Comment (roccoblues) [reply util.go:97]: fair enough, let's keep it then.

Comment (skaes) [reply util.go:97]: So, rename yes or no?

Comment (roccoblues) [reply util.go:97]: no, let's keep it like it is.

File: middleware.go (hunk 3 above)
Comment (skaes) [reply middleware.go:49]: It's a user error. I think it's helpful to send the info to logjam and that the program does not silently die. Whether the request actually makes it to logjam  is a different question. I'm really missing an at_exit handler concept.

Comment (skaes) [reply middleware.go:49]: And at the very least it is useful for local development. We might even want to add a stacktrace here. What do you think?

Comment (skaes) [reply middleware.go:49]: Never mind the last point, we can think about that later.

**Pull request merged**